libs/server/Storage/Session/MainStore/VectorStoreOps.cs

```diff
@@ -0,0 +1,279 @@
+        {
+            parseState.InitializeWithArgument(ArgSlice.FromPinnedSpan(key.AsReadOnlySpan()));
+
+            var input = new RawStringInput(RespCommand.VSIM, ref parseState);
+
+            Span<byte> indexSpan = stackalloc byte[VectorManager.IndexSizeBytes];
+
+            using (vectorManager.ReadVectorIndex(this, ref key, ref input, indexSpan, out var status))
+            {
+                if (status != GarnetStatus.OK)
+                {
+                    result = VectorManagerResult.Invalid;
+                    outputIdFormat = VectorIdFormat.Invalid;
+                    return status;
+                }
+
+                result = vectorManager.ElementSimilarity(indexSpan, element, count, delta, searchExplorationFactor, filter, maxFilteringEffort, includeAttributes, ref outputIds, out outputIdFormat, ref outputDistances, ref outputAttributes);
+                return GarnetStatus.OK;
+            }
+        }
+
+        /// <summary>
+        /// Get the approximate vector associated with an element, after (approximately) reversing any transformation.
+        /// </summary>
+        [SkipLocalsInit]
+        public unsafe GarnetStatus VectorSetEmbedding(SpanByte key, ReadOnlySpan<byte> element, ref SpanByteAndMemory outputDistances)
+        {
+            parseState.InitializeWithArgument(ArgSlice.FromPinnedSpan(key.AsReadOnlySpan()));
+
+            var input = new RawStringInput(RespCommand.VEMB, ref parseState);
+
+            Span<byte> indexSpan = stackalloc byte[VectorManager.IndexSizeBytes];
+
+            using (vectorManager.ReadVectorIndex(this, ref key, ref input, indexSpan, out var status))
+            {
+                if (status != GarnetStatus.OK)
+                {
+                    return status;
+                }
+
+                if (!vectorManager.TryGetEmbedding(indexSpan, element, ref outputDistances))
+                {
+                    return GarnetStatus.NOTFOUND;
+                }
+
+                return GarnetStatus.OK;
+            }
+        }
+
+        [SkipLocalsInit]
+        internal unsafe GarnetStatus VectorSetDimensions(SpanByte key, out int dimensions)
+        {
+            parseState.InitializeWithArgument(ArgSlice.FromPinnedSpan(key.AsReadOnlySpan()));
+
+            var input = new RawStringInput(RespCommand.VDIM, ref parseState);
+
+            Span<byte> indexSpan = stackalloc byte[VectorManager.IndexSizeBytes];
+
+            using (vectorManager.ReadVectorIndex(this, ref key, ref input, indexSpan, out var status))
+            {
+                if (status != GarnetStatus.OK)
+                {
+                    dimensions = 0;
+                    return status;
+                }
+
+                // No need to recreate, all of this data is available to Garnet alone
+
+                // After a successful read we add the vector while holding a shared lock
+                // That lock prevents deletion, but everything else can proceed in parallel
```

Comment:
I'm a bit confused, does this comment apply for VDIM? In VectorManager.ReadIndex we're only interpreting the index metadata raw bytes, right? No locks as well, correct?

Reply:
Copy/pasta error - it's not about adding, it's about reading metadata.

We still need to acquire the lock because that's how we (generally) validate the that we're working on a Vector Set.  We could optimize that away, but VDIM is marginal enough (and read locks fast enough) that it hasn't been done.

I will correct comment and make a note about some missing (ie. `WRONGTYPE`) behavior.